### PR TITLE
Refactor ProviderApprovalController to use rpc and publicConfigStore

### LINKED ISSUE #6030: web3.eth.coinbase is still null when metamask is re-logged in.
**Describe the bug**
web3.eth.coinbase is still null when metamask is re-logged in.

**To Reproduce**
1. Login to metamask
2. Go to console in browser
3. Type _console.log(window.web3.eth.coinbase);_ //return correct coinbase
4. Log out metamask
5. Type _console.log(window.web3.eth.coinbase);_ //return null
6. Login to metamask
7. Type _console.log(window.web3.eth.coinbase);_ //still return null

**Browser details (please complete the following information):**
 - OS: OS
 - Browser Chome
 - MetaMask Version: 5.3.0

### LINKED ISSUE #6263: MM 6.2.1 sends invalid JSONRPC requests to private network
Metamask 6.2.1 seems to be broken in my dev environment. I'm running parity 2.3.5 with a private network, and it seems to be sending the following request:

    $ curl http://localhost:8545 -H 'Content-Type: application/json' -d '{"method":"net_version","params":[],"beta":true,"id":2790068417}'
    {"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request"},"id":2790068417}

It seems to have replaced `"jsonrpc":"2.0"` with `"beta":true` ?

This shows up in the browser console as:

    MetaMask - RPC Error: Internal JSON-RPC error. 
    {code: -32603, message: "Internal JSON-RPC error."}

### LINKED ISSUE #6462: Silent Block Tracker optimization removed
This PR added an always-on `latest` event on the blockTracker, removing our "silent BlockTracker" optizations
https://github.com/MetaMask/metamask-extension/commit/8bf866699202f70d35308c68f1cbe2d7608ec64c#diff-dde3377567c2957d676403ae2e0f9605L8

### PR DESCRIPTION
**Pull request by kumavis**
### STATUS: Ready

I noticed the `provider.enable` and `eth_requestAccounts` feature was not implemented using our existing patterns of streams/dnode/publicConfigStore. It ended up needing lots of custom plumbing and state management in inpage, contentscript, and background. Here I reworked the feature to take advantage of our existing patterns.

### Notable Changes:
- background's `publicConfigStore` is now created per-domain and the state of exposed accounts are managed there
- the deprecated `provider.enable` is now just a wrapper around `eth_requestAccounts`
- `ProviderApprovalController` exposes middleware for handling `eth_requestAccounts` requests
- Contentscript has been rewired a bit. Removed the old, dead `PingPongStream`. We now use the multiplexer to split out the individual channels and only forward the allowed channels across inpage<->background. Contentscript now runs a dnode api for exposing `getSiteMetadata` (unless im mistaken, only contentscript exposes the dapp dom). I previously tried to solve this problem via `SiteMetadataController` but it had lifecycle issues for GC of the metadata. 
- removed `web3.currentProvider._metamask.isApproved` as it was not distinct from `isEnabled`
- `web3.currentProvider._metamask.isEnabled` now returns a promise

### QA checks:
- requesting accounts via rpc
- requesting accounts via provider.enable
- requesting accounts via provider.enable after its enabled
- request UI uses image and title from page
- reset allowed sites should hide addresses without restart

Fixes #6462 (Silent BlockTracker optimization removed)
Fixes #6263 (Invalid JSON-RPC requests on testnet)
Fixes #6030 (web3.eth.coinbase null after re-login)

### CODE REVIEW

**Comment by kumavis** [comment]
QA async eth_accounts

File: app/scripts/controllers/provider-approval.js [hunk 1]
```diff
@@ -14,38 +16,40 @@ class ProviderApprovalController {
...
+   *
+   * @param {object} opts - opts for the middleware contains the origin for the middleware
+   */
+  createMiddleware ({ origin, getSiteMetadata }) {
+    return createAsyncMiddleware(async (req, res, next) => {
+      // only handle requestAccounts
+      if (req.method !== 'eth_requestAccounts') return next()
+      // register the provider request
+      const metadata = await getSiteMetadata(origin)
+      console.log('metadata', metadata)
```
Comment (danfinlay) [inline app/scripts/controllers/provider-approval.js]: Log!

**Review by kumavis — COMMENTED** [review]
### Refactor Provider Approval Controller TODOs
- [x] check async eth_accounts
- [x] de-platform createBlockTracker
- [x] remove platform.sendMessage

**Comment by kumavis** [comment]
noticed we were adding a listener for latest block in `createBlockTracker` - this removes all the optimizations we had around silent block-tracker
^ @frankiebee

**Review by bitpshr — COMMENTED** [review]
This code looks really clean and is a huge improvement over the hand-rolled `postMessage` implementation.  After some testing, I noticed that the convenience methods at `ethereum._metamask` all return `undefined` after page reload until the extension is opened by the user; after the extension is manually opened, all convenience methods return (or resolve to) boolean values as expected when called thereafter.

Everything else worked perfectly.

**Comment by kumavis** [comment]
@bitpshr ok I'll address the `_metamask` api behavior

**Comment by kumavis** [comment]
i should also address this https://github.com/MetaMask/metamask-extension/issues/6263

**Comment by kumavis** [comment]
ok `_metamask` api should be working, though changed
`isEnabled` is now async, `isApproved` is removed

**Comment by kumavis** [comment]
@bitpshr if you could leave a GH review, that would be much appreciated

**Comment by kumavis** [comment]
status: Ready for review : )

**Comment by bitpshr** [comment]
@kumavis: Why did the signature for `isEnabled` change and why did `isApproved` get removed? Several dapps use both of these functions, so modifying them or removing them will be a breaking change downstream.

**Comment by kumavis** [comment]
@bitpshr I changed `isEnabled` to be async because the value is not available immediately on page load (requires background to push the value to the page. this is not a new requirement.)
I removed `isApproved` because it didn't seem different from `isEnabled`. 

I didn't really think twice about modifying these because we have this warning

> ethereum._metamask exposes methods that have not been standardized yet. [...] and may be removed from MetaMask in the future.

you said
> Several dapps use both of these functions

can you tell me more about this? Are these methods documented somewhere?

**Comment by bitpshr** [comment]
@kumavis: The methods exposed at `ethereum._metamask` are proprietary to MetaMask's 1102 implementation. Since not all dapp browser wallets have a notion of being locked or support caching, both `isUnlocked` and `isApproved` never made the final browser-agnostic EIP. Lots of downstream developers made a _ton_ of noise about this functionality: many dapps rely on knowing if MetaMask is unlocked, and many more wanted to know if an approval pop up _would_ show if `ethereum.enable()` is called without actually calling it, which is what `isApproved` does.

TLDR; Relying on those convenience methods is brittle, ties a dapp to MetaMask-specific functionality, and is all-around bad form. Still, dapps use them despite all the warnings and `_metamask`-specific namespace we added them to. Removing them now would amount to a breaking change. This may be fine, but we should document this and maybe write a blog post so dapps know what's coming.

Also, re: `isApproved`: the background push is fine as long as it happens as fast as account population was happening in the pre-1102 MetaMask. Even though this isn't technically synchronously available at page load, it's fast enough. This is how `isApproved` currently works on `develop`.

**Comment by kumavis** [comment]
@bitpshr
> Removing them now would amount to a breaking change

undocumented apis behind an underscore-prefixed name **must** be considered safe to change or we can never change anything. This is why the warning is there. These APIs haven't been around for very long anyways.
 
I don't understand what you mean by "caching" as stated here
>  Since not all dapp browser wallets [...] support caching [...]

> Also, re: isApproved: the background push is fine as long as it happens as fast as account population 

they are populated by the same mechanism : ) obs-store via background

I still dont understand the difference between `isEnabled` and `isApproved`, besides `isApproved` is async.

per my current understanding:
we'll defined "appoval state" as if the domain can e.g. read the accounts (via approval or disabling privacy mode)
`isApproved` asyncly returns the approval state
`isEnabled` syncly returns the approval state with a potential false negative if invoked too early

does this sound correct?

**Comment by bitpshr** [comment]
@kumavis:

`isApprved` returns a `Promise` that resolves to `true` or `false` indicating if there is a cached approval for the current domain. A cached approval means this domain has been approved by the user in the past, so an approval popup will not show when `ethereum.enable` is called.

`isEnabled` returns `true` or `false` indicating if the current domain is or is not approved by the user for the current session.

So, `isApproved` indicates past, saved approvals, and `isEnabled` indicates approval for the current session. They are different.

**Comment by kumavis** [comment]
@bitpshr ok
but why are they different? What is the purpose of a notion of session here?

> so an approval popup will not show when ethereum.enable is called.

so you're saying that if we have previously approved a domain in a previous session:
- `isEnabled()` should return false
- `ethereum.enable()` will not show a confirmation window
- `isEnabled()` should now return true?

that flow doesnt really make sense to me

**Comment by bitpshr** [comment]
@kumavis: We cache approvals to reduce the number of popups a user is shown when browsing a site, especially a multi-page site. Imagine using GitHub and having to respond to an account-request popup every time you hit a new page.

The flow you outlined above is correct; dapps should always call `ethereum.enable()`, even if they have a cached approval. I originally didn't want to expose `ethereum._metamask.isApproved` at all since I felt that dapps shouldn't know if the user previously approved them or not, but this proved to be a tough UX pill to swallow downstream.

**Comment by kumavis** [comment]
@bitpshr thanks for the clarification

> isEnabled always starts as false and accounts are not populated regardless of cached approval.

do you consider this behavior important? I see no issue with `isEnabled` being true and accounts populated automatically if approved in a previous session. In that situation `enable()` can just be a noop that returns the accounts.

**Comment by bitpshr** [comment]
@kumavis: I think that's fine. FWIW, that's how Mist implemented cached approvals: `isEnable` became unnecessary since users could just check for populated accounts.

As long as we maintain that method (for now) so dapps calling it don't break, auto-populating accounts for cached dapps works for me. That's one step closer to getting rid of one of these pesky convenience methods, so that's a win in my mind.

**Comment by kumavis** [comment]
@bitpshr ok then this is ready to merge : )

**Review by bitpshr — APPROVED** [review]
Looks good to me. The local test cases I ran worked as expected, too. This is huge work @kumavis, thanks for making this uniform with the rest of our dapp <-> bg communication.
